services/lsst-dev.rst

```diff
@@ -116,106 +116,139 @@ Select Appropriate Developer Tools
+The following Software Collections are currently available:
 
-If you prefer, you can enable the `Red Hat Developer Toolset`_ version 3 (``devtoolset-3``) which has been pre-installed.
-This provides an updated toolchain, including GCC 4.9.2.
-If you enable this, you must not use the shared stack; instead you must build your own version of the entire stack using this toolchain.
+================ ================================================
+Name             Description
+================ ================================================
+``devtoolset-3`` Updated compiler toolchain providing GCC 4.9.2.
+``devtoolset-4`` Updated compiler toolchain providing GCC 5.3.1.
+``devtoolset-6`` Updated compiler toolchain providing GCC 6.3.1.
+``devtoolset-7`` Updated compiler toolchain providing GCC 7.1.1.
+``git19``        The `Git`_ version control system version 1.9.4.
+================ ================================================
 
-Enable and test ``devtoolset-3`` using the ``scl`` command as follows (replacing ``bash`` with your shell of choice if necessary):
+To enable a particular Software Collection use the ``scl`` command. For example:
 
 .. prompt:: bash $ auto
 
-   $ scl enable devtoolset-3 bash
+   $ scl enable devtoolset-6 bash
    $ gcc --version
-   gcc (GCC) 4.9.2 20150212 (Red Hat 4.9.2-6)
-   Copyright (C) 2014 Free Software Foundation, Inc.
+   gcc (GCC) 6.3.1 20170216 (Red Hat 6.3.1-3)
+   Copyright (C) 2016 Free Software Foundation, Inc.
    This is free software; see the source for copying conditions.  There is NO
    warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 
-The Developer Toolset includes version 1.9.3 of the `Git`_ version control system.
-If you prefer the (slightly) more recent version 1.9.4, you may also wish to enable the ``git19`` package.
-This may be done at the same time as enabling ``devtoolset-3``.
+.. warning::
 
-.. prompt:: bash
-
-   scl enable devtoolset-3 git19 bash
+   Code compiled by different versions of GCC may not be compatible: it is generally better to stick to a particular toolchain for a given project.
+   In particular, if you are using a :ref:`shared stack <lsst-dev-loadlsst>` you *must* use the matching toolchain.
 
-You may wish to automatically enable ``devtoolset-3`` every time you log in to ``lsst-dev01`` by adding it to your shell initialization files.
-For example, try adding the following to :file:`~/.profile`:
+You may wish to automatically enable a particular software collection every time you log in to ``lsst-dev01`` and other LSST systems.
+Take care if you do this: it's easy to accidentally to either start recursively spawning shells and run out of resources or lock yourself out of machines which don't have the particular collection you're interested in installed.
+If you are using `Bash`_ — the default shell on ``lsst-dev01`` — try placing the following at the end of :file:`~/.bash_profile` and customising the list of ``desired_scls``.
```

Comment:
@parejkoj @mfisherlevine I think the below provides the sort of instructions you are looking for. Doing this in a way which doesn't break if `scl` is missing, if `devtoolset-n` is missing, if `devtoolset-n` is already activated, which enables the user to activate both `git19` and `devtoolset-n` at the same time, etc etc gets complicated, and I'm no shell wizard — could you let me know if this does the trick?

Reply:
That looks pretty good, but putting it at the end of my `.bash_profile` (which loads `.bashrc` which has `source /software/lsstsw/stack/loadLSST.bash` at the end) does result in the following redundancy on login:

```
Note: use devtoolset-6 in conjunction with this stack.
Enabling git19 devtoolset-6.
Note: use devtoolset-6 in conjunction with this stack.
```

Reply:
Makes sense: your `.bash_profile` is sourcing `.bashrc` which sources `loadLSST.sh`, then `exec`ing another copy of `bash`, which once again sources `.bashrc` and hence `loadLSST.sh`. Probably fine to just add the `loadLSST.bash` line to your `.bash_profile` before the `exec`, I think.